repo/mk/core.misc.channels.promote.mk

```diff
@@ -10,29 +18,17 @@ perform-promotion/%:
 	$(eval TAG_COMMIT := $(shell $(GIT) rev-list -n1 $(TAG)))
 	$(eval GIT_REMOTE=$(shell $(GIT) config branch.master.remote))
 
-	$(ECHO) "$(SF_RELEASE_CHANNELS)" | $(GREP) -q -e "\(^\|\s\)$(PROMOTE_CHANNEL)\(\s\|$$\)" || { \
+	$(ECHO) "$(SF_PROMOTE_CHANNELS)" | $(GREP) -q -e "\(^\|\s\)$(PROMOTE_CHANNEL)\(\s\|$$\)" || { \
 		$(ECHO_ERR) "$(PROMOTE_CHANNEL) is not a known channel."; \
+		$(ECHO_INFO) "Known channels: $(SF_PROMOTE_CHANNELS)"; \
 		exit 1; \
 	}
 	$(GIT) tag --list | $(GREP) -q "^$(TAG)$$" || { \
 		$(ECHO_ERR) "$(TAG) is not a tag."; \
 		exit 1; \
 	}
 	$(GIT) fetch
-	$(ECHO)
-	$(ECHO_INFO) "Changes ready to be promoted:"
-	$(ECHO)
-	$(GIT) --no-pager log \
-		--color \
-		--graph \
-		--date=short \
-		--pretty=format:"%h %ad %s" \
-		--no-decorate \
-		$(GIT_REMOTE)/$(PROMOTE_BRANCH)..$(TAG_COMMIT) | \
-		$(GREP) --color -i -E "^|break" || true
-	$(ECHO)
-	$(ECHO_INFO) "Breaking changes ready to be promoted:"
-	$(ECHO)
+	$(MAKE) show-channel-status
```

Comment:
so these are not equivalent.
before it was showing "changes to be promoted", and after it is showing "the 20 most recent commits" *in the current branch* (regardless of what promote-branch and tag are)
🤔

Reply:
You're right. I felt that the "changes to be promoted" diff didnt give enough info when it came to the releases repo since it only has release commits. It felt to me like an overview of the channels' current status was better for the promoting person to confirm that they did indeed promote to the correct version. 

For example it was decided that the latest canary release was good enough that we should promote stable channel to the same release. When I make the promotion of stable to 1.1.3, before I confirm it I get an overview to actually see that I promoted the correct version since I can see that canary is currently on 1.1.3

I do however agree though that showing X amount of arbitrary commits wouldnt cover all bases, since there could be many channels or releases in between promotions

Just an example of the difference that I feel can be useful: 

```
~/git/tobii/merlin-controller-electron: internal-track ● ❯ make promote-channel/test/v1.0.1
make: Entering directory '/Users/jban/git/tobii/merlin-controller-electron'
              > users-jban-git-tobii-merlin-controller-electron-cloning-git-github-com-tobiipro-merlin-controller-electron-releases-git-into-releases
10:28:51 [DO  ] Cloning git@github.com:tobiipro/merlin-controller-electron-releases.git into releases...
releases
10:28:54 [DONE] in 0 seconds
              < users-jban-git-tobii-merlin-controller-electron-cloning-git-github-com-tobiipro-merlin-controller-electron-releases-git-into-releases

make[1]: Entering directory '/Users/jban/git/tobii/merlin-controller-electron/releases'
10:28:56 [INFO] Changes ready to be promoted:

* 79b439a 2020-07-06 v1.0.1
* 37979db 2020-07-05 v1.0.0
* bded84a 2020-06-26 v0.9.2
* 83fa77e 2020-06-18 v0.9.1

10:28:57 [INFO] Breaking changes ready to be promoted:


[Q   ] Still want to promote v1.0.1 to test?
       Press ENTER to Continue.
       Press Ctrl+C to Cancel.


```

vs 

```
~/git/tobii/merlin-controller-electron: internal-track ● ❯ make promote-channel/test/v1.0.2
make: Entering directory '/Users/jban/git/tobii/merlin-controller-electron'
              > users-jban-git-tobii-merlin-controller-electron-cloning-git-github-com-tobiipro-merlin-controller-electron-releases-git-into-releases
11:03:28 [DO  ] Cloning git@github.com:tobiipro/merlin-controller-electron-releases.git into releases...
releases
11:03:31 [DONE] in 0 seconds
              < users-jban-git-tobii-merlin-controller-electron-cloning-git-github-com-tobiipro-merlin-controller-electron-releases-git-into-releases

make[1]: Entering directory '/Users/jban/git/tobii/merlin-controller-electron/releases'
make[2]: Entering directory '/Users/jban/git/tobii/merlin-controller-electron/releases'

11:03:33 [INFO] Current channel status:

* 5224550 (HEAD -> master, tag: v1.1.4, origin/master, origin/HEAD) v1.1.4
* 175839f (tag: v1.1.3, origin/stable, origin/canary) v1.1.3
* 7c9ddd5 (tag: v1.1.2) v1.1.2
* 528fe36 (tag: v1.1.0) v1.1.0
* 3306b6f (tag: v1.0.3, origin/merlin) v1.0.3
* b0c0a0c (tag: v1.0.2) v1.0.2
* 79b439a (tag: v1.0.1) v1.0.1
* 37979db (tag: v1.0.0) v1.0.0
* bded84a (tag: v0.9.2) v0.9.2
* 83fa77e (tag: v0.9.1) v0.9.1
* 0ab6064 (tag: v0.9.0, origin/test) v0.9.0
* 0e1f95e (tag: v0.8.3) v0.8.3
* b529529 (tag: v0.8.2) v0.8.2
* 06307ce (tag: v0.8.1) v0.8.1
* 8e04214 (tag: v0.8.0) v0.8.0
* 89fddf2 (tag: v0.7.2) v0.7.2
* 9e9e073 push
* d1e2561 (tag: v0.7.1) v0.7.1
* ab83aad push
* d2a40ed (tag: v0.7.0) v0.7.0
make[2]: Leaving directory '/Users/jban/git/tobii/merlin-controller-electron/releases'

[Q   ] Still want to promote v1.0.2 to test?
       Press ENTER to Continue.
       Press Ctrl+C to Cancel.

```

Reply:
I'm with you now. You want to show a history between the current channel's release and the latest release (master) along with annotations of where each channel (branch) is at. Makes sense 👏 

And `make view-channels` is a similar, but since it doesn't have a starting point (I guess ideally it would be the oldest-updated channel) we go with 20 commits back in time.

PS: for fun, this is what gives you the the oldest-updated channel

```Makefile
view-channels:
	$(eval SF_PROMOTE_CHANNELS_REMOTE_REFS := $(patsubst %,refs/heads/$(GIT_REMOTE)/%,$(SF_PROMOTE_CHANNELS)))
	$(eval OLDEST_CHANNEL := $(shell \
		$(GIT) for-each-ref --format="%(committerdate:unix) %(refname:short)" $(SF_PROMOTE_CHANNELS_REMOTE_REFS) | \
		$(SORT) -k1 | \
		$(HEAD) -1 | \
		$(CUT) -d" " -f2))
	...
```